controller/pkg/packet/packet.go

```diff
@@ -439,33 +458,90 @@ func (p *Packet) TCPDataAttach(tcpOptions []byte, tcpData []byte) (err error) {
+	return p.IPHdr.sourceAddress.String() + ":" + p.IPHdr.destinationAddress.String() + ":" + strconv.Itoa(int(p.SourcePort())) + ":" + strconv.Itoa(int(p.DestPort()))
 }
 
 // L4ReverseFlowHash calculate a hash string based on the 4-tuple by reversing source and destination information
 func (p *Packet) L4ReverseFlowHash() string {
-	return p.DestinationAddress.String() + ":" + p.SourceAddress.String() + ":" + strconv.Itoa(int(p.DestinationPort)) + ":" + strconv.Itoa(int(p.SourcePort))
+	return p.IPHdr.destinationAddress.String() + ":" + p.IPHdr.sourceAddress.String() + ":" + strconv.Itoa(int(p.DestPort())) + ":" + strconv.Itoa(int(p.SourcePort()))
 }
 
 // SourcePortHash calculates a hash based on dest ip/port for net packet and src ip/port for app packet.
 func (p *Packet) SourcePortHash(stage uint64) string {
 	if stage == PacketTypeNetwork {
-		return p.DestinationAddress.String() + ":" + strconv.Itoa(int(p.DestinationPort))
+		return p.IPHdr.destinationAddress.String() + ":" + strconv.Itoa(int(p.DestPort()))
 	}
-	return p.SourceAddress.String() + ":" + strconv.Itoa(int(p.SourcePort))
+
+	return p.IPHdr.sourceAddress.String() + ":" + strconv.Itoa(int(p.SourcePort()))
 }
 
 // ID returns the IP ID of the packet
 func (p *Packet) ID() string {
-	return strconv.Itoa(int(p.ipID))
+	return strconv.Itoa(int(p.IPHdr.ipID))
 }
 
 //TCPOptionLength returns the length of tcpoptions
 func (p *Packet) TCPOptionLength() int {
-	return len(p.tcpOptions)
+	return len(p.TCPHdr.tcpOptions)
 }
 
 //TCPDataLength -- returns the length of tcp options
 func (p *Packet) TCPDataLength() int {
-	return len(p.tcpData)
+	return len(p.TCPHdr.tcpData)
+}
+
+//SourcePort -- returns the appropriate source port
+func (p *Packet) SourcePort() uint16 {
+	if p.IPHdr.ipProto == IPProtocolTCP {
+		return p.TCPHdr.sourcePort
+	}
+
+	return p.UDPHdr.sourcePort
+}
+
+//DestPort -- returns the appropriate destination port
+func (p *Packet) DestPort() uint16 {
+	if p.IPHdr.ipProto == IPProtocolTCP {
+		return p.TCPHdr.destinationPort
+	}
+
+	return p.UDPHdr.destinationPort
+}
+
+//SourceAddress returns the source IP
+func (p *Packet) SourceAddress() net.IP {
+	return p.IPHdr.sourceAddress
+}
+
+//DestinationAddress returns the destination address
+func (p *Packet) DestinationAddress() net.IP {
+	return p.IPHdr.destinationAddress
+}
+
+//IPProto returns the L4 protocol
+func (p *Packet) IPProto() uint8 {
+	return p.IPHdr.ipProto
+}
+
+//IPTotalLen returns the total length of the packet
+func (p *Packet) IPTotalLen() uint16 {
+	return p.IPHdr.ipTotalLength
+}
+
+//IPHeaderLen returns the ip header length
+func (p *Packet) IPHeaderLen() uint8 {
+	return p.IPHdr.ipHeaderLen
+}
+
+//TestGetTCPPacket is used by other test code when they need to create a packet
+func TestGetTCPPacket(srcIP, dstIP net.IP, srcPort, dstPort uint16) *Packet {
```

Comment:
Functions written in _test files are not exported. datapath test had to populate a packet and since we are not exposing the data structures we need this as part of the main file. I am not happy about this either.